src/nvim/os/fileio.c

```diff
@@ -284,14 +284,17 @@ ptrdiff_t file_read(FileDescriptor *const fp, char *const ret_buf,
   char *buf = ret_buf;
   size_t read_remaining = size;
   RBuffer *const rv = fp->rv;
+  bool called_read = false;
   while (read_remaining) {
     const size_t rv_size = rbuffer_size(rv);
     if (rv_size > 0) {
       const size_t rsize = rbuffer_read(rv, buf, MIN(rv_size, read_remaining));
       buf += rsize;
       read_remaining -= rsize;
     }
-    if (fp->eof) {
+    if (fp->eof
+        // Allow only at most one os_read[v] call.
+        || called_read && fp->non_blocking) {
```

Comment:
we usually prefer to be explicit about operator precedence:

    || (called_read && fp->non_blocking)) {